src/ast/writing/writer.ts

```diff
@@ -1,14 +1,61 @@
 import { ASTNode, ASTNodeConstructor } from "../ast_node";
 import { YulNode } from "../implementation/statement/inline_assembly";
-import { Statement } from "../implementation/statement/statement";
 import { SourceFormatter } from "./formatter";
 
 export interface YulNodeWriter {
     write(node: YulNode, writer: YulWriter): string;
 }
 
-export interface ASTNodeWriter {
-    write(node: ASTNode, writer: ASTWriter, cache: Map<ASTNode, string>): string;
+export type SrcRangeMap = Map<ASTNode, [number, number]>;
+export type DescArgs = Array<string | ASTNode | undefined | null>;
+/**
+ * The `SrcDesc` is the intermediate description for source that is generated by
+ * `ASTNodeWriter`s. It has a tree-like structure, defined by the following grammar:
+ *
+ * SrcDesc ::= (string | [ASTNode, SrcDesc])*
+ *
+ * Essentially its a tree of strings, with some ASTNode attached to it at given locations.
+ *
+ * For example having `[ASTNode#5, <SrcDescX>]` in the tree, indicates that the
+ * source generated by <SrcDescX> corresponds precisely to `ASTNode#5` in the
+ * source map.
+ */
+export type SrcDesc = Array<string | [ASTNode, any[]]>;
+
+/**
+ * Base class for all ASTNodeWriters. Child classes are responsible for
+ * generating a `SrcDesc` for every node.
+ */
+export abstract class ASTNodeWriter {
+    /**
+     * Generate a `SrcDesc` for a given node, but without adding the
+     * node itself to the tree.
+     *
+     * @param node - node
+     * @param writer - writer
+     */
+    abstract writeInt(node: ASTNode, writer: ASTWriter): SrcDesc;
+    /**
+     * Add the node to the descrioption generated by `writeInt`, any additional 'wrappings' and return it. Usually `write` is responsible
+     * for adding semicolons and documentation, since these are not generally part of source mappings.
+     *
+     * For example given this source:
+     * `a = 1;`
+     *
+     * `ExpressionStatement.writeInt` would return the following desc:
+     *
+     * [[Assignment#3, [[Identifier#1, ["a"]], " = ", [Literal#2, ["1"]] ]]]
```

Comment:
The most concerning part here is the amount of object allocations for the deeply nested trees. Unsure how much memory it will consume.

Reply:
Its true that compared to the other scheme we are allocating a lot of small arrays here. On the other hand we are omitting the string scanning that was done in ASTSourceMapComputer before, and getting more precise and fine-grained results. So we are trading memory for speed and precision. Perhaps we need to add some memory profiling to track how much each stage consumes?